WIP: Support remote signing over RPC

guggero opened the pull request:
Fixes https://github.com/lightningnetwork/lnd/issues/5486.
Also took some ideas from https://github.com/lightningnetwork/lnd/issues/3944, maybe we can close that as well.

This PR makes it possible to run an instance of `lnd` in purely watch-only mode (no private key material in the wallet whatsoever). All signing or ECDH requests are then forwarded to a secondary `lnd` instance (or, in theory, any software that implements the necessary RPC interfaces) for processing.

It is important that the secondary (the remote signing instance with the private) keys is _NOT_ used as a wallet. The keys and addresses are kept in sync between the watch-only and the signing instance, so manually generating addresses or deriving keys on the signing instance will lead to unexpected errors on the watch-only instance!


TODO for removing WIP state:
 - Documentation
 - Fix itest, add more test cases

Roasbeef commented on config.go:
```diff
@@ -1542,6 +1546,35 @@ func (c *Config) ImplementationConfig() ImplementationCfg {
+	// If we're using a remote signer, we still need the base wallet as a
+	// watch-only source of chain and address data. But we don't need any
+	// private key material in that btcwallet base wallet.
+	if c.RemoteSigner.Enable {
```
So I think this is close, but it should still be one "level" higher. So rather than us checking this portion of the config here, we'd instead do it within `cmd/lnd`. The difference is that in that instances, `lnd` doesn't need to handle the initialization, and instead just accepts the fully initialized interfaces. This way, things can evolve outside of lnd (via wrapper binaries kinda like `litd`) without needing to change anything on the bsae `lnd` level. 

In practice, this means more config parsing being moved into `cmd/lnd`.

Roasbeef commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -0,0 +1,433 @@
+	// for pre-0.13.0 channels are attempted to be restored. In that case
+	// it is recommended to restore the channels using a node with the full
+	// seed available.
+	return nil, ErrRemoteSigningPrivateKeyNotAvailable
```
🙅‍♂️

Roasbeef commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -286,6 +286,16 @@ message AddrRequest {
     string account = 1;
+
+    /*
+    The type of address to derive.
```
Ah totally forgot about this, guess we've been defaulting to P2WKH, will be useful post taproot as well.

Roasbeef commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -73,14 +83,104 @@ func NewRPCKeyRing(watchOnlyWallet keychain.SecretKeyRing,
+			"signer instance: %v", err)
+	}
+
+	localAddr, err := r.WalletController.NewAddress(
```
In the future, I think we could extra this wallet sync logic to a higher layer, but likely good enough for it to live here now.

Roasbeef commented on lnwallet/rpcwallet/rpcwallet.go:
```diff
@@ -73,14 +83,104 @@ func NewRPCKeyRing(watchOnlyWallet keychain.SecretKeyRing,
+	dryRun bool) (*waddrmgr.AccountProperties, []btcutil.Address,
+	[]btcutil.Address, error) {
+
+	// TODO(guggero): Implement.
```
I would imagine this has already happened as part of the initial wallet init?

Roasbeef commented on lnrpc/walletunlocker.proto:
```diff
@@ -188,6 +197,63 @@ message InitWalletResponse {
+    required by some hardware wallets for proper identification and signing. The
+    bytes must be in big-endian order.
+    */
+    bytes master_key_fingerprint = 2;
```
Is there a strong requirement that this is on this level? In that, can it go one layer below within the `WatchOnlyAccount` struct?

Roasbeef reviewed (changes requested):
The final component falls into place! 🍾

Happy to see all the incremental steps we've taken over the past year or so resulted in this PR primary being some refactoring and additional layers of indirection. 

Initial pass complete, main comment is the additional layer of indirection needed to allow the "primary" interfaces to be _passed into_ the `lnd.Main` method rather than `lnd` initializing the concrete version of the interfaces _within_ itself. Also see the msgs sent offline re some sample directions.

guggero commented:
I've extracted all the refactoring commits into https://github.com/lightningnetwork/lnd/pull/5708, will address the rest of the comments later.

Roasbeef commented on lntest/itest/lnd_remote_signer_test.go:
```diff
@@ -0,0 +1,247 @@
+		Purpose:  keychain.BIP0043Purpose,
+		CoinType: harnessNetParams.HDCoinType,
+		Account:  uint32(keychain.KeyFamilyTowerID),
+		Xpub: "tpubDDXFHr67Ro2tgE89V8ZdgMytC2Jq1iT9ttGhdzR1X7haQJNBmX" +
```
Seems we need to extend the existing RPCs to also have a way to obtain these xpubs. 

Tagging this here based on our offline convo

Roasbeef replied:
Could extend `lncli wallet accounts list` to include this information

Roasbeef commented:
So I think we want something like this: 
```diff
diff --git a/keychain/btcwallet.go b/keychain/btcwallet.go
index 8d5dab3bd..3110fc960 100644
--- a/keychain/btcwallet.go
+++ b/keychain/btcwallet.go
@@ -64,7 +64,7 @@ type BtcWalletKeyRing struct {
 //
 // NOTE: The passed waddrmgr.Manager MUST be unlocked in order for the keychain
 // to function.
-func NewBtcWalletKeyRing(w *wallet.Wallet, coinType uint32) SecretKeyRing {
+func NewBtcWalletKeyRing(w *wallet.Wallet, coinType uint32) (SecretKeyRing, error) {
 	// Construct the key scope that will be used within the waddrmgr to
 	// create an HD chain for deriving all of our required keys. A different
 	// scope is used for each specific coin type.
@@ -73,10 +73,33 @@ func NewBtcWalletKeyRing(w *wallet.Wallet, coinType uint32) SecretKeyRing {
 		Coin:    coinType,
 	}
 
-	return &BtcWalletKeyRing{
+	// TODO(xxx): create all keyscopes here
+
+	wallet := &BtcWalletKeyRing{
 		wallet:        w,
 		chainKeyScope: chainKeyScope,
 	}
+
+	for _, keyFam := range VersionZeroKeyFamilies {
+		db := wallet.wallet.Database()
+		err := walletdb.Update(db, func(tx walletdb.ReadWriteTx) error {
+			addrmgrNs := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+
+			scope, err := wallet.keyScope()
+			if err != nil {
+				return err
+			}
+
+			// If the account doesn't exist, then we may need to create it
+			// for the first time in order to derive the keys that we
+			// require.
+			return wallet.createAccountIfNotExists(addrmgrNs, keyFam, scope)
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return wallet, nil
 }
 
 // keyScope attempts to return the key scope that we'll use to derive all of
diff --git a/keychain/derivation.go b/keychain/derivation.go
index b22d8f70f..15ac954e8 100644
--- a/keychain/derivation.go
+++ b/keychain/derivation.go
@@ -105,6 +105,21 @@ const (
 	KeyFamilyTowerID KeyFamily = 9
 )
 
+// VersionZeroKeyFamilies is a slice of all the known key families for first
+// version of the key derivation schema defined in this package.
+var VersionZeroKeyFamilies = []KeyFamily{
+	KeyFamilyMultiSig,
+	KeyFamilyRevocationBase,
+	KeyFamilyHtlcBase,
+	KeyFamilyPaymentBase,
+	KeyFamilyDelayBase,
+	KeyFamilyRevocationRoot,
+	KeyFamilyNodeKey,
+	KeyFamilyStaticBackup,
+	KeyFamilyTowerSession,
+	KeyFamilyTowerID,
+}
+
 // KeyLocator is a two-tuple that can be used to derive *any* key that has ever
 // been used under the key derivation mechanisms described in this file.
 // Version 0 of our key derivation schema uses the following BIP43-like

```

Roasbeef commented:
The other half (lets `lncli wallet accounts list`) display all the pubkeys: 
```
diff --git a/lnwallet/btcwallet/btcwallet.go b/lnwallet/btcwallet/btcwallet.go
index 1db9e0b18..496ec1723 100644
--- a/lnwallet/btcwallet/btcwallet.go
+++ b/lnwallet/btcwallet/btcwallet.go
@@ -552,6 +552,18 @@ func (b *BtcWallet) ListAccounts(name string,
 			account := account
 			res = append(res, &account.AccountProperties)
 		}
+
+		accounts, err = b.wallet.Accounts(waddrmgr.KeyScope{
+			Purpose: keychain.BIP0043Purpose,
+			Coin:    b.cfg.CoinType,
+		})
+		if err != nil {
+			return nil, err
+		}
+		for _, account := range accounts.Accounts {
+			account := account
+			res = append(res, &account.AccountProperties)
+		}
 	}
 
 	return res, nil
```

guggero commented:
Rebased and updated with the above suggestions.
Now the only thing needed is a new command (perhaps `lncli create-watchonly`?) that takes the output of `lncli wallet accounts list` and converts it into the format that is needed for the `InitWallet` RPC.

weltitob commented:
@Roasbeef  @guggero  can I find an example remote signer implementation repo? just displaying the conenction? I always get a timeout trying to recreate the setup
